packages/shared/src/lib/core/isc/classes/isc-magic-contracts.class.ts

```diff
@@ -0,0 +1,135 @@
+import { ISC_MAGIC_CONTRACT_ADDRESS } from '@core/layer-2'
+import { IIscChain } from '@core/network'
+import { Contract } from 'web3'
+import { ISC_MAGIC_CONTRACT_SANDBOX_ABI } from '../abis'
+import { IscAssets, IscDict, IscHName, IscL1Address, IscSendMetadata, IscSendOptions } from '../types'
+import { evmAddressToAgentId } from '@core/layer-2/helpers'
+import { ISC_MAGIC_CONTRACT_ACCOUNTS_ABI } from '../abis/isc-magic-contract-accounts.abi'
+
+export interface MethodObject {
+    encodeABI: () => string
+    estimateGas: () => Promise<number>
+}
+
+interface ISendableIscMagicContractSandboxMethods {
+    send: {
+        targetAddress: IscL1Address
+        assets: IscAssets
+        adjustMinimumStorageDeposit: boolean
+        sendMetadata: IscSendMetadata
+        sendOptions: IscSendOptions
+    }
+    call: {
+        contractName: IscHName
+        entryPoint: IscHName
+        params: IscDict
+        allowance: IscAssets
+    }
+}
+
+class IscMagicContractSandbox {
+    private _contract: Contract<typeof ISC_MAGIC_CONTRACT_SANDBOX_ABI>
+
+    constructor(iscChain: IIscChain, contractAddress: string) {
+        this._contract = iscChain.getContract(ISC_MAGIC_CONTRACT_SANDBOX_ABI, contractAddress)
+    }
+
+    encode<K extends keyof ISendableIscMagicContractSandboxMethods>(
+        method: K,
+        params: ISendableIscMagicContractSandboxMethods[K]
+    ): string {
+        // Use a type assertion to let TypeScript know the type of this[method]
+        const methodFunction = this[method] as unknown as (
+            params: ISendableIscMagicContractSandboxMethods[K]
+        ) => MethodObject
+        return methodFunction(params).encodeABI()
+    }
+
+    estimateGas<K extends keyof ISendableIscMagicContractSandboxMethods>(
+        method: K,
+        params: ISendableIscMagicContractSandboxMethods[K]
+    ): Promise<number> {
+        // Use a type assertion to let TypeScript know the type of this[method]
+        const methodFunction = this[method] as unknown as (
+            params: ISendableIscMagicContractSandboxMethods[K]
+        ) => MethodObject
```

Comment:
the reason for my comment below is this assignment as an anonymous function. With this, the context of `this` changes.